lib/net.js

```diff
@@ -761,13 +743,10 @@ Socket.prototype._writeGeneric = function(writev, data, encoding, cb) {
 
   this._unrefTimer();
 
-  let req;
   if (writev)
-    req = writevGeneric(this, data, cb);
+    writevGeneric(this, data, cb);
   else
-    req = writeGeneric(this, data, encoding, cb);
-  if (req.async)
-    this[kLastWriteQueueSize] = req.bytes;
+    writeGeneric(this, data, encoding, cb);
```

Comment:
`writableLength` won't get subtracted until the write actually completes.

Reply:
This is my understanding: assume there is nothing buffered so `writableLength` is zero and there is a write in progress, `bufferSize` should report the number of bytes that are currently being written and not zero.

Reply:
I don't think writableLength is zero here, it becomes non zero as soon as you do write() and won't become zero again unless all writes have finished

Reply:
Ok, can you check when this was originally added? The original commit might shed some light.

Reply:
Perhaps at that time `writableLength` wasn't working like this? I don't know.

Reply:
https://github.com/nodejs/node/pull/17650

Reply:
unclear...

Reply:
The only explanation I can think of is if the callback is called before the write has actually completed... which would be a bit weird.

Reply:
Right, I don’t think that that’s what happens (but I guess it could be checked easily enough through verifying that the write queue size is 0 when we get the callback from libuv?)

Reply:
Seems like `handle.writeQueueSize` is *not* zero before invoking the callback. Which is kind of confusing.

Reply:
@addaleax the following assertion will fail when running the test suite:

```js
diff --git a/lib/internal/stream_base_commons.js b/lib/internal/stream_base_commons.js
index fdf540a47e..4c7049bf51 100644
--- a/lib/internal/stream_base_commons.js
+++ b/lib/internal/stream_base_commons.js
@@ -30,6 +30,7 @@ const {
 } = require('internal/timers');
 const { isUint8Array } = require('internal/util/types');
 const { clearTimeout } = require('timers');
+const assert = require('assert');
 
 const kMaybeDestroy = Symbol('kMaybeDestroy');
 const kUpdateTimer = Symbol('kUpdateTimer');
@@ -104,6 +105,8 @@ function onWriteComplete(status) {
   stream[kUpdateTimer]();
   stream[kAfterAsyncWrite](this);
 
+  assert(this.handle.writeQueueSize === 0);
+
   if (typeof this.callback === 'function')
     this.callback(null);
 }
```